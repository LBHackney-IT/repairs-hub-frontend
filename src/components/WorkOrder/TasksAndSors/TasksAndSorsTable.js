import PropTypes from 'prop-types'
import TasksAndSorsRow from './TasksAndSorsRow'
import { Table, THead, TBody, TR, TH, TD } from '../../Layout/Table'
import Collapsible from 'src/components/Layout/Collapsible/Collapsible'
import { formatDateTime } from '../../../utils/time'
import { calculateTotal } from '../../../utils/helpers/calculations'

const TasksAndSorsTable = ({
  latestTasksAndSors,
  originalTasksAndSors,
  tabName,
  tasksWereUpdated,
}) => {
  const buildTable = (tasks, isOriginal = false) => {
    return (
      <>
        <THead>
          <TR className="lbh-body">
            <TH scope="col">SOR</TH>
            <TH scope="col">Description</TH>
            <TH scope="col">Quantity (est.)</TH>
            <TH scope="col" type="numeric">
              Unit cost
            </TH>
            <TH scope="col" type="numeric">
              Cost (est.)
            </TH>
            <TH scope="col" type="numeric">
              Total SMV
            </TH>
          </TR>
        </THead>
        <TBody>
          {tasks.map((entry, index) => (
            <TasksAndSorsRow
              key={index}
              index={index}
              isOriginal={isOriginal}
              taskQuantity={
                isOriginal ? entry.originalQuantity : entry.quantity
              }
              {...entry}
            />
          ))}
          <TR
            index={
              isOriginal
                ? originalTasksAndSors.length
                : latestTasksAndSors.length
            }
            className="lbh-body-s"
          >
            <TD>{}</TD>
            <TD>{}</TD>
            <TD>{}</TD>
            <TD type="numeric">
              <strong>Total</strong>
            </TD>
            <TD type="numeric">
              <strong>
                Â£
                {calculateTotal(
                  tasks,
                  'cost',
                  `${isOriginal ? 'originalQuantity' : 'quantity'}`
                ).toFixed(2)}
              </strong>
            </TD>
            <TD type="numeric">
              <strong>
                {calculateTotal(
                  tasks,
                  `${isOriginal ? 'originalQuantity' : 'quantity'}`,
                  'standardMinuteValue'
                )}
              </strong>
            </TD>
          </TR>
        </TBody>
      </>
    )
  }

  const displayReadableDate = (tasks, isOriginal = false) => {
    if (isOriginal || !tasksWereUpdated) {
      return `Added on ${formatDateTime(tasks[0].dateAdded)}`
    } else {
      let latestUpdatedTask = tasks.reduce((a, b) => {
        return new Date(a.dateUpdated) > new Date(b.dateUpdated) ? a : b
      })
      return `Added on ${formatDateTime(latestUpdatedTask.dateUpdated)}`
    }
  }

  return (
    <>
      <h2 className="lbh-heading-h2">{tabName}</h2>

      <h4 className="lbh-heading-h4">
        Latest Tasks and SORs{' '}
        {latestTasksAndSors && latestTasksAndSors.length > 0
          ? displayReadableDate(latestTasksAndSors)
          : ''}
      </h4>
      <Table className="govuk-!-margin-top-5 latest-tasks-and-sors-table">
        {buildTable(latestTasksAndSors)}
      </Table>
      <br />
      <br />
      {tasksWereUpdated ? (
        <Collapsible
          heading={`Original Tasks and SORs 
        ${
          originalTasksAndSors && originalTasksAndSors.length > 0
            ? displayReadableDate(originalTasksAndSors, true)
            : ''
        }`}
          startClosed={true}
        >
          <Table className="govuk-!-margin-top-5 original-tasks-and-sors-table">
            {buildTable(originalTasksAndSors, true)}
          </Table>
        </Collapsible>
      ) : (
        ''
      )}
    </>
  )
}

TasksAndSorsTable.propTypes = {
  tabName: PropTypes.string.isRequired,
  originalTasksAndSors: PropTypes.array.isRequired,
  tasksWereUpdated: PropTypes.bool.isRequired,
  latestTasksAndSors: PropTypes.arrayOf(
    PropTypes.shape({
      code: PropTypes.string,
      description: PropTypes.string,
      dateAdded: PropTypes.instanceOf(Date),
      original: PropTypes.bool,
      quantity: PropTypes.number,
      cost: PropTypes.number,
      status: PropTypes.string,
    })
  ).isRequired,
}

export default TasksAndSorsTable
