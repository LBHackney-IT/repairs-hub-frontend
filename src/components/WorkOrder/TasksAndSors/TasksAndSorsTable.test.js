import { render } from '@testing-library/react'
import TasksAndSorsTable from './TasksAndSorsTable'

describe('TasksAndSorsTable component', () => {
  const props = {
    tabName: 'Tasks and SORs',
    originalTasksAndSors: [
      {
        code: 'DES5R006',
        description: 'Urgent call outs',
        dateAdded: new Date('2021-02-03T11:33:35.757339'),
        original: true,
        originalQuantity: 1,
        quantity: 2,
        cost: 20.15,
        standardMinuteValue: 10,
      },
    ],
    latestTasksAndSors: [
      {
        code: 'DES5R006',
        description: 'Urgent call outs',
        dateAdded: new Date('2021-02-03T11:33:35.757339'),
        original: true,
        originalQuantity: 1,
        quantity: 2,
        cost: 20.15,
        standardMinuteValue: 25,
      },
      {
        code: 'DES5R005',
        description: 'Normal call outs',
        dateAdded: new Date('2021-02-03T11:33:35.814437'),
        original: false,
        originalQuantity: null,
        quantity: 4,
        cost: 1.4,
        standardMinuteValue: 30,
      },
      {
        code: 'DES5R013',
        description: 'Inspect additional sec entrance',
        dateAdded: new Date('2021-02-03T11:33:35.799566'),
        original: false,
        originalQuantity: null,
        quantity: 5,
        cost: 7.31,
        standardMinuteValue: 15,
      },
    ],
  }

  it('should render properly', () => {
    const { asFragment } = render(
      <TasksAndSorsTable
        originalTasksAndSors={props.originalTasksAndSors}
        latestTasksAndSors={props.latestTasksAndSors}
        tabName={props.tabName}
      />
    )
    expect(asFragment()).toMatchSnapshot()
  })
})
