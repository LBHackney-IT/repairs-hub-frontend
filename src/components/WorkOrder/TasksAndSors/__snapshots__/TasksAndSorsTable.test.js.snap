// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TasksAndSorsTable component should render properly 1`] = `
<DocumentFragment>
  <h2
    class="lbh-heading-h2"
  >
    Tasks and SORs
  </h2>
  <h4
    class="lbh-heading-h4"
  >
    Latest Tasks and SORs Added on 3 Feb 2021, 11:33
  </h4>
  <table
    class="govuk-table lbh-table govuk-!-margin-top-5 latest-tasks-and-sors-table"
  >
    <thead
      class="govuk-table__head"
    >
      <tr
        class="govuk-table__row lbh-body"
      >
        <th
          class="govuk-table__header"
          scope="col"
        >
          SOR
        </th>
        <th
          class="govuk-table__header"
          scope="col"
        >
          Description
        </th>
        <th
          class="govuk-table__header"
          scope="col"
        >
          Quantity (est.)
        </th>
        <th
          class="govuk-table__header govuk-table__header--numeric"
          scope="col"
        >
          Unit cost
        </th>
        <th
          class="govuk-table__header govuk-table__header--numeric"
          scope="col"
        >
          Cost (est.)
        </th>
        <th
          class="govuk-table__header govuk-table__header--numeric"
          scope="col"
        >
          Total SMV
        </th>
      </tr>
    </thead>
    <tbody
      class="govuk-table__body"
    >
      <tr
        class="govuk-table__row lbh-body-s"
        data-row-id="0"
      >
        <td
          class="govuk-table__cell"
        >
          DES5R006
        </td>
        <td
          class="govuk-table__cell"
        >
          Urgent call outs
        </td>
        <td
          class="govuk-table__cell"
        >
          2
        </td>
        <td
          class="govuk-table__cell govuk-table__cell--numeric"
        >
          £20.15
        </td>
        <td
          class="govuk-table__cell govuk-table__cell--numeric"
        >
          £40.30
        </td>
        <td
          class="govuk-table__cell govuk-table__cell--numeric"
        >
           50
        </td>
      </tr>
      <tr
        class="govuk-table__row lbh-body-s"
        data-row-id="1"
      >
        <td
          class="govuk-table__cell"
        >
          DES5R005
        </td>
        <td
          class="govuk-table__cell"
        >
          Normal call outs
        </td>
        <td
          class="govuk-table__cell"
        >
          4
        </td>
        <td
          class="govuk-table__cell govuk-table__cell--numeric"
        >
          £1.4
        </td>
        <td
          class="govuk-table__cell govuk-table__cell--numeric"
        >
          £5.60
        </td>
        <td
          class="govuk-table__cell govuk-table__cell--numeric"
        >
           120
        </td>
      </tr>
      <tr
        class="govuk-table__row lbh-body-s"
        data-row-id="2"
      >
        <td
          class="govuk-table__cell"
        >
          DES5R013
        </td>
        <td
          class="govuk-table__cell"
        >
          Inspect additional sec entrance
        </td>
        <td
          class="govuk-table__cell"
        >
          5
        </td>
        <td
          class="govuk-table__cell govuk-table__cell--numeric"
        >
          £7.31
        </td>
        <td
          class="govuk-table__cell govuk-table__cell--numeric"
        >
          £36.55
        </td>
        <td
          class="govuk-table__cell govuk-table__cell--numeric"
        >
           75
        </td>
      </tr>
      <tr
        class="govuk-table__row lbh-body-s"
      >
        <td
          class="govuk-table__cell"
        />
        <td
          class="govuk-table__cell"
        />
        <td
          class="govuk-table__cell"
        />
        <td
          class="govuk-table__cell govuk-table__cell--numeric"
        >
          <strong>
            Total
          </strong>
        </td>
        <td
          class="govuk-table__cell govuk-table__cell--numeric"
        >
          <strong>
            £ 82.45
          </strong>
        </td>
        <td
          class="govuk-table__cell govuk-table__cell--numeric"
        >
          <strong>
            245
          </strong>
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;
